base/applications/dxdiag/system.c

```diff
@@ -225,32 +227,73 @@ InitializeSystemPage(HWND hwndDlg)
     if (GetLocaleInfo(LOCALE_SYSTEM_DEFAULT,LOCALE_SLANGUAGE , szTime, sizeof(szTime) / sizeof(WCHAR)))
         SendDlgItemMessageW(hwndDlg, IDC_STATIC_LANG, WM_SETTEXT, 0, (LPARAM)szTime);
 
+    /* prepare SMBIOS data */
+    SMBiosBuf = LoadSMBiosData(DmiStrings);
+
     /* set system manufacturer */
     szTime[0] = L'\0';
-    if (GetRegValue(HKEY_LOCAL_MACHINE, L"Hardware\\Description\\System\\BIOS", L"SystemManufacturer", REG_SZ, szTime, sizeof(szTime)))
+    Length = 0;
+    if (SMBiosBuf)
+    {
+        Length = GetSMBiosStringW(DmiStrings[SYS_VENDOR], szTime, sizeof(szTime) / sizeof(WCHAR), FALSE);
+    }
+    if (Length == 0)
+    {
+        Length = GetRegValue(HKEY_LOCAL_MACHINE, L"Hardware\\Description\\System\\BIOS", L"SystemManufacturer", REG_SZ, szTime, sizeof(szTime));
+    }
+    if (Length > 0)
     {
         szTime[199] = L'\0';
         SendDlgItemMessageW(hwndDlg, IDC_STATIC_MANU, WM_SETTEXT, 0, (LPARAM)szTime);
     }
 
     /* set motherboard model */
     szTime[0] = L'\0';
-    if (GetRegValue(HKEY_LOCAL_MACHINE, L"Hardware\\Description\\System\\BIOS", L"SystemProductName", REG_SZ, szTime, sizeof(szTime)))
+    Length = 0;
+    if (SMBiosBuf)
+    {
+        Length = GetSMBiosStringW(DmiStrings[SYS_PRODUCT], szTime, sizeof(szTime) / sizeof(WCHAR), FALSE);
+    }
+    if (Length == 0)
+    {
+        Length = GetRegValue(HKEY_LOCAL_MACHINE, L"Hardware\\Description\\System\\BIOS", L"SystemProductName", REG_SZ, szTime, sizeof(szTime));
+    }
+    if (Length > 0)
     {
         SendDlgItemMessageW(hwndDlg, IDC_STATIC_MODEL, WM_SETTEXT, 0, (LPARAM)szTime);
     }
 
     /* set bios model */
     szTime[0] = L'\0';
-    if (GetRegValue(HKEY_LOCAL_MACHINE, L"Hardware\\Description\\System\\BIOS", L"BIOSVendor", REG_SZ, szTime, sizeof(szTime)))
+    Length = 0;
+    if (SMBiosBuf)
+    {
+        Length = GetSMBiosStringW(DmiStrings[BIOS_VENDOR], szTime, sizeof(szTime) / sizeof(WCHAR), TRUE);
+    }
+    if (Length == 0)
+    {
+        Length = GetRegValue(HKEY_LOCAL_MACHINE, L"Hardware\\Description\\System\\BIOS", L"BIOSVendor", REG_SZ, szTime, sizeof(szTime));
+    }
+    if (Length > 0)
     {
         DWORD Index;
         DWORD StrLength = (sizeof(szTime) / sizeof(WCHAR));
 
         Index = wcslen(szTime);
+        if(Index + 1 < sizeof(szTime))
+            szTime[Index++] = L' ';
         StrLength -= Index;
 
-        if (GetRegValue(HKEY_LOCAL_MACHINE, L"Hardware\\Description\\System\\BIOS", L"BIOSReleaseDate", REG_SZ, &szTime[Index], StrLength))
+        Length = 0;
+        if (SMBiosBuf)
+        {
+            Length = GetSMBiosStringW(DmiStrings[BIOS_DATE], &szTime[Index], StrLength / sizeof(WCHAR), TRUE);
+        }
+        if (Length == 0)
+        {
+            Length = GetRegValue(HKEY_LOCAL_MACHINE, L"Hardware\\Description\\System\\BIOS", L"BIOSReleaseDate", REG_SZ, &szTime[Index], StrLength);
+        }
+        if (Length > 0)
```

Comment:
We see this pattern 4 times, maybe introduce a small helper function?

Reply:
Added helper function :)